File: 2022/Sources/advent2022/Day24.swift
struct Day24 : Day {

    let sampleInput = [
        "#.######",
        "#>>.<^<#",
        "#.<..<<#",
        "#>v.><>#",
        "#<^v^^>#",
        "######.#",
    ]

    enum Move : UInt8, CaseIterable {
        case Wait = 0, Up = 0x1, Right = 0x2, Down = 0x4, Left = 0x8
        var dx :Int { self == .Left ? -1 : self == .Right ? 1 : 0 }
        var dy :Int { self == .Up ? -1 : self == .Down ? 1 : 0 }
    }

    struct Valley {
        let width :Int, height :Int
        let entry :Int, exit :Int
        let cells :[UInt8]

        init (_ input :[String]) {
            width = input[0].count - 2
            height = input.count - 2
            entry = input.first!.firstIndexInt(of: ".")!-1
            exit = input.last!.firstIndexInt(of: ".")!-1
            let cellMap :[Character: Move] = [".": .Wait, ">": .Right, "^": .Up, "v": .Down, "<": .Left]
            var cells = [UInt8](repeating: 0, count: width*height)
            for ii in input.indices {
                let yy = ii-1
                if yy < 0 || yy >= height { continue }
                let line = input[ii]
                let idxS = line.index(line.startIndex, offsetBy: 1)
                let idxE = line.index(line.startIndex, offsetBy: line.count-1)
                cells.replaceSubrange(yy*width..<(yy+1)*width, with: line[idxS..<idxE].map({ cellMap[$0]!.rawValue }))
            }
            self.cells = cells
        }

        func blow (_ cells :[UInt8]) -> [UInt8] {
            var ncells = [UInt8](repeating: 0, count: cells.count)
            for yy in 0 ..< height {
                for xx in 0 ..< width {
                    let cell = cells[yy*width+xx]
                    for dir in Move.allCases {
                        if cell & dir.rawValue != 0 {
                            let nx = (xx + dir.dx + width) % width
                            let ny = (yy + dir.dy + height) % height
                            ncells[ny*width+nx] |= dir.rawValue
                        }
                    }
                }
            }
            return ncells
        }

        func startToExit (_ phases :Int) -> Int {
            route(phases, [(entry, -1), (exit, height)]) }
        func startToExitToStartToExit (_ phases :Int) -> Int {
            route(phases, [(entry, -1), (exit, height), (entry, -1), (exit, height)]) }

        func route (_ phases :Int, _ route :[(Int, Int)]) -> Int {
            var steps = 0, ncells = cells
            for ii in route.indices {
                if ii == route.count-1 { break }
                let (sx, sy) = route[ii], (ex, ey) = route[ii+1]
                let (nncells, psteps) = path(phases, ncells, sx, sy, ex, ey)
                steps += psteps
                ncells = nncells
            }
            return steps
        }

        func path (_ phases :Int, _ cells :[UInt8], _ sx :Int, _ sy :Int, _ ex :Int, _ ey :Int) -> ([UInt8], Int) {
            func phaseCoord (_ x :Int, _ y :Int, _ phase :Int) -> Int { phase * 10000 + x * 100 + y }
            var bests = [Int: Int]()
            func h (_ x :Int, _ y :Int) -> Int { (abs(x-ex) + abs(y-ey)) }
            var nexts = [(cells, sx, sy, 0, h(sx, sy))]
            while nexts.count > 0 {
                let (cells, x, y, steps, _) = nexts.removeFirst()
                let ncells = blow(cells), nsteps = steps+1, phase = nsteps % phases
                for dir in Move.allCases {
                    let nx = x + dir.dx, ny = y + dir.dy
                    if nx == ex && ny == ey { return (ncells, nsteps) }
                    if nx == sx && ny == sy && dir == .Wait {} // allow waiting at start even though it's OOB
                    else if nx < 0 || nx >= width || ny < 0 || ny >= height || ncells[ny*width+nx] != 0 { continue }
                    let ncost = nsteps+h(nx, ny), pcoord = phaseCoord(nx, ny, phase)
                    if let bcost = bests[pcoord], bcost <= ncost { continue }
                    bests[pcoord] = ncost
                    nexts.insertSorted((ncells, nx, ny, nsteps, ncost), { (a, b) in a.4 < b.4 })
                }
            }
            return (cells, 0)
        }
    }

    func part1 () throws -> String { String(Valley(try readInput(24)).startToExit(20)) }
    func part2 () throws -> String { String(Valley(try readInput(24)).startToExitToStartToExit(20)) }
}

File: 2022/scala/src/main/scala/Day24.scala
object Day24 extends Day(24):

  enum Move (val mask :Byte, val dx :Int, val dy :Int):
    case Wait  extends Move(0,    0,  0)
    case Up    extends Move(0x1,  0, -1)
    case Right extends Move(0x2,  1,  0)
    case Down  extends Move(0x4,  0,  1)
    case Left  extends Move(0x8, -1,  0)
  import Move._

  case class Pos (x :Int, y :Int):
    def move (dir :Move) = Pos(x+dir.dx, y+dir.dy)
    def in (width :Int, height :Int) = x >= 0 && x < width && y >= 0 && y < height
    def idx (width :Int) = y*width + x
    def h (end :Pos) = Math.abs(x-end.x) + Math.abs(y-end.y)
    def phaseCoord (phase :Int) = phase*10000 + x*100 + y

  class Valley (input :Seq[String]):
    val width = input(0).size-2
    val height = input.size-2
    val entry = input(0).indexOf('.')-1
    val exit = input.last.indexOf('.')-1
    val cells = {
      val cellMap = Map('.' -> Wait, '>' -> Right, '^' -> Up, 'v' -> Down, '<' -> Left)
      var cells = new Array[Byte](width*height)
      for (ii <- input.indices ; yy = ii-1 if yy >= 0 && yy < height)
        val line = input(ii)
        val row = line.substring(1, line.size-1).map(c => cellMap(c).mask).toArray
        System.arraycopy(row, 0, cells, yy*width, width)
      cells
    }

    def blow (cells :Array[Byte]) =
      val ncells = new Array[Byte](cells.length)
      for (idx <- ncells.indices)
        val yy = idx / width ; val xx = idx % width ; val cell = cells(idx)
        for (dir <- Move.values if (cell & dir.mask) != 0)
          val nx = (xx + dir.dx + width) % width
          val ny = (yy + dir.dy + height) % height
          val nidx = ny*width+nx
          ncells(nidx) = (ncells(nidx) | dir.mask).toByte
      ncells

    def startToExit (phases :Int) =
      route(phases, Seq(Pos(entry, -1), Pos(exit, height)))
    def startToExitToStartToExit (phases :Int) =
      route(phases, Seq(Pos(entry, -1), Pos(exit, height), Pos(entry, -1), Pos(exit, height)))

    def route (phases :Int, route :Seq[Pos]) =
      var steps = 0 ; var ncells = cells
      for (ii <- route.indices.dropRight(1))
        val start = route(ii) ; val end = route(ii+1)
        val (nncells, psteps) = path(phases, ncells, start, end)
        steps += psteps
        ncells = nncells
      steps

    def path (phases :Int, cells :Array[Byte], start :Pos, end :Pos) :(Array[Byte], Int) =
      var bests = Map[Int, Int]()
      val first = (cells, start, 0, start.h(end))
      var nexts = collection.mutable.PriorityQueue(first)(Ordering.by(-_._4))
      while nexts.size > 0 do
        val (cells, pos, steps, _) = nexts.dequeue()
        if pos == end then return (cells, steps)
        val ncells = blow(cells) ; val nsteps = steps+1 ; val phase = nsteps % phases
        for (dir <- Move.values)
          val npos = pos.move(dir)
          if (npos == start && dir == Wait) || (npos == end) || // allow start/end even though OOB
             (npos.in(width, height) && ncells(npos.idx(width)) == 0) then
            val ncost = nsteps + npos.h(end) ; val npc = npos.phaseCoord(phase)
            if bests.getOrElse(npc, ncost+1) > ncost then
              bests += (npc -> ncost)
              nexts += ((ncells, npos, nsteps, ncost))
      (cells, 0)

  override def answer1 (input :Seq[String]) = Valley(input).startToExit(20)
  override def answer2 (input :Seq[String]) = Valley(input).startToExitToStartToExit(20)
